envoy/server/factory_context.h

```diff
@@ -194,6 +194,27 @@ class ServerFactoryContext : public virtual CommonFactoryContext {
   virtual bool healthCheckFailed() const PURE;
 };
 
+/**
+ * Factory context for configuration parsing and initiation. This context contains server factory
+ * context reference and a validation visitor from uncertain context (server, listener, cluster or
+ * others).
+ */
+class ConfigFactoryContext {
+public:
+  virtual ~ConfigFactoryContext() = default;
+
+  /**
+   * @return ServerFactoryContext which lifetime is no shorter than the server and provides
+   *         access to the server's resources.
+   */
+  virtual ServerFactoryContext& getServerFactoryContext() const PURE;
+
+  /**
+   * @return ProtobufMessage::ValidationVisitor& validation visitor for configuration messages.
+   */
+  virtual ProtobufMessage::ValidationVisitor& messageValidationVisitor() const PURE;
+};
```

Comment:
Formatter extensions maybe used in both server scope and listener scope. We should use different for message validation visitors in different scopes.

See https://github.com/envoyproxy/envoy/issues/31073 for more context.